src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1910,6 +1911,25 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
                         }
                     }
 
+                    CastKind::Hide => {
+                        let predicates = match ty.sty {
+                            ty::Opaque(def_id, substs) => {
+                                let bounds = tcx.predicates_of(def_id);
+                                let result = bounds.instantiate(tcx, substs);
+                                // TODO: do I need to normalize associated types here somehow.
+                                // as is done in coercion.rs?
+                                result
+                            }
+                            _ => bug!(),
```

Comment:
Will the opaque type always be at the top level? I imagine there might be something like

```rust
let x: (impl Debug, impl Debug) = (22, 44);
```

I think what we want to do here instead is to use [`instantiate_opaque_types`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/infer/struct.InferCtxt.html#method.instantiate_opaque_types) to replace the `ty::Opaque` with type variables. This will give us back a type that we can unify with the type of the value being casted. It will also give us back the predicates to prove that you can prove using `prove_predicates` as below.